### DISCUSSION TITLE
Using Protocol::HTTP in a non-Async-http application

### DISCUSSION THREAD

**Discussion by Forthoney** [ioquatix/ioquatix#30] [Q&A] [answered]
In my search for a good, pure-ruby HTTP Parser, I ran into Protocol::HTTP. I think it's perfect for my use case, but I'm having some trouble actually integrating it into my project. Pretty much all of the examples across the Protocol::HTTP libraries assume it's being used in the context of Async-http, making it difficult to figure out how things may work without Async-http.

I'm currently stuck [here](https://github.com/Forthoney/mooro/blob/protocol/lib/mooro/plugin/http.rb). My application currently hangs at [L23](https://github.com/Forthoney/mooro/blob/6a32fe5cfbcd04db6ed1600e211d3dad481ec4d4/lib/mooro/plugin/http.rb#L23). At this point, the response has already been properly formed, but it fails to actually deliver it to the client (i.e. my browser). When I turn the server off via SIGINT, the response is displayed.

The code is essentially identical to `Async::HTTP::Protocol::HTTP1::Server` because that's the only other use of this gem in this use-case I could find. My assumption is something I have in there doesn't work without Async, but I'm also not very familiar with the `async` gem to debug this.

* Note that the `Adapter.initialize`'s signature has been changed to also take in a logger object. This was done to get around the global object sharing limitation of Ractors

**Comment by ioquatix** [comment]
This is a reasonable use case and it's great that the gem/code is useful to you.

I don't see anything wrong with your code, but `read_request` also shouldn't hang unless there is some problem with the request. Are you able to give me a simple repro script so I can run it on my end?

**Reply by ioquatix** [reply]
Here is a working plain-IO server: https://github.com/socketry/protocol-http1/blob/main/examples/http1/server.rb

**Reply by Forthoney** [reply]
Here's the sample I was running through `bin/console`
```ruby
class TestServer < Mooro::Server
  include Mooro::Plugin::HTTP
end

server = TestServer.new(4)
server.start
```

**Reply by Forthoney** [reply]
> Here is a working plain-IO server: https://github.com/socketry/protocol-http1/blob/main/examples/http1/server.rb

Thank you for the example! I think I have the ergonomics of reading and writing to the connection correctly. My gut feeling is that the issue lies in how I handle persistent/keep-alive connections, since once I remove the `while request = next_request` loop and just do `request = next_request`, the message is actually delivered to the browser. Perhaps I write the body upon serving a single request, but am missing a bit of code that **flushes** the socket?

**Reply by ioquatix** [reply]
Okay, it sounds like I should add persistent connection handling to my example and maybe we'd find the issue?

**Reply by Forthoney** [reply]
> Okay, it sounds like I should add persistent connection handling to my example and maybe we'd find the issue?

Yes this would be amazing. Thank you so much for the help!

**Reply by Forthoney** [reply]
@ioquatix Don't mean to nag you but let me know if there's any updates!

**Reply by ioquatix** [reply, marked as answer]
Can you check https://github.com/socketry/protocol-http1/commit/4b46cffd7bafec5ddeb13ca9f998160f3014093f and let me know how you get on.

Make sure you read the request body if any before reading the next request.

**Reply by Forthoney** [reply]
Yes this works great. Thank you again for the help and the great library!

**Comment by Forthoney** [comment]
Hi @ioquatix , I think I may have found an issue with the `server.rb` example. It works fine with `curl`, but when testing the connection through a browser, it doesn't recognize when the browser disconnected. This is an edge case likely due to the browsers doing some magic behind the scenes, I think it's worth getting right considering many people like to test things out on their browser.
 Here's a screen recording of what happens. 

https://github.com/ioquatix/ioquatix/assets/85617794/2e35bbca-1c36-4601-bb4f-7b231f916f70

**Reply by Forthoney** [reply]
Ahh I found the issue - it is indeed due to browser magic lol. [Apparently](https://news.ycombinator.com/item?id=26565134) browsers hang on to connections for a specific time period (~2 minutes for firefox) even after a window is closed. Apologies for the confusion!

**Reply by ioquatix** [reply]
I didn't know that either! Nice find. Actually, browsers can share HTTP/2 connections between tabs too IIRC.
